Add TCPRoute and UDPRoute Support for L4 Load Balancing

Skcey opened the pull request:
### Proposed changes
Problem:
NGF currently lacks support for TCPRoute and UDPRoute resources, which are essential for managing Layer 4 (TCP/UDP) traffic via the Gateway API.

Solution: 
This PR adds basic support for TCPRoute and UDPRoute to NGF, enabling Layer 4 load balancing. Key implementations include:
1. Added controllers to watch and process TCPRoute/UDPRoute resources, following the pattern used for HTTPRoute/TLSRoute.
2. Implemented route construction logic in the state graph to resolve listeners, backend services, and reference grants for L4 routes.
3. Extended the NGINX stream configuration generator to create upstream groups and server blocks for TCP/UDP traffic, mapping routes to their backend services.

Testing: 
Packaged into an image and tested in my environment.

Closes #3687

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [ ✔️] I have read the [CONTRIBUTING](https://github.com/nginx/nginx-gateway-fabric/blob/main/CONTRIBUTING.md) doc
- [✔️ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [ ✔️] I have rebased my branch onto main
- [ ✔️] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### Release notes

If this PR introduces a change that affects users and needs to be mentioned in the [release notes](../blob/main/CHANGELOG.md),
please add a brief note that summarizes the change.

<!-- If this PR does not require a release note, you can just write NONE in the release-note block below. -->

```release-note

```

Skcey commented:
I have hereby read the F5 CLA and agree to its terms

Skcey commented:
recheck

shaun-nx commented on internal/controller/provisioner/objects.go:
```diff
@@ -43,6 +43,11 @@ const (
 	defaultInitialDelaySeconds = int32(3)
 )
 
+type PortInfo struct {
+	Port     int32
+	Protocol corev1.Protocol
+}
+
```
Just curious, why did you make a struct for this?
From the way this is being used, it feels like you can reference Protocol on its own.

For example, on line `144`, we can define `ports := make(map[int32]corev1.Protocol)`
Line `155` then becomes `ports[int32(listener.Port)] = protocol`

Then the loop on line 473 becomes this

```go
for port, protocol := range ports {
  servicePort := corev1.ServicePort{
	  Name:       fmt.Sprintf("port-%d", port),
	  Port:       port,
	  TargetPort: intstr.FromInt32(port),
	  Protocol:   protocol,
```

Would love to know what you think though. Do please tell me if I'm overlooking anything.

shaun-nx commented on internal/controller/state/dataplane/configuration.go:
```diff
@@ -168,6 +172,79 @@ func buildPassthroughServers(gateway *graph.Gateway) []Layer4VirtualServer {
+			continue
+		}
+
+		if len(l.L4Routes) > 1 {
+			fmt.Printf(
+				"WARN: Listener %s has %d TCPRoutes, which is not supported. Skipping.",
+				l.Name,
+				len(l.L4Routes),
+			)
+			continue
+		}
```
If this warning occurs, will the user see this information in their resource status?

sjberman replied:
We shouldn't be using print statements in the code anyway. We have a logger to use. Regardless, we should be doing validation and setting status conditions before we even get to this point in the code. It would be when we build the graph, which is the initial step. If we're building the configuration here, this validation should have already occurred.

ciarams87 commented on internal/controller/state/graph/tcproute.go:
```diff
@@ -0,0 +1,125 @@
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	"github.com/nginx/nginx-gateway-fabric/internal/controller/state/conditions"
```
```suggestion
	"github.com/nginx/nginx-gateway-fabric/v2/internal/controller/state/conditions"
```

ciarams87 commented on internal/controller/state/graph/udproute.go:
```diff
@@ -0,0 +1,125 @@
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	"github.com/nginx/nginx-gateway-fabric/internal/controller/state/conditions"
```
```suggestion
	"github.com/nginx/nginx-gateway-fabric/v2/internal/controller/state/conditions"
```

ciarams87 commented on internal/controller/state/dataplane/configuration.go:
```diff
@@ -242,6 +319,137 @@ func buildStreamUpstreams(
+
+			allowedAddressType := getAllowedAddressType(ipFamily)
+
+			eps, err := serviceResolver.Resolve(ctx, br.SvcNsName, br.ServicePort, allowedAddressType)
```
This function call needs to be updated as it's missing the logger parameter - see the pattern in buildUpstreams:

`eps, err := svcResolver.Resolve(ctx, logger, br.SvcNsName, br.ServicePort, allowedAddressType)`

ciarams87 commented on internal/controller/state/dataplane/configuration.go:
```diff
@@ -242,6 +319,137 @@ func buildStreamUpstreams(
+
+			allowedAddressType := getAllowedAddressType(ipFamily)
+
+			eps, err := serviceResolver.Resolve(ctx, br.SvcNsName, br.ServicePort, allowedAddressType)
```
As above - this function call needs to be updated as it's missing the logger parameter - see the pattern in buildUpstreams:

`eps, err := svcResolver.Resolve(ctx, logger, br.SvcNsName, br.ServicePort, allowedAddressType)`

ciarams87 reviewed (changes requested):
Hi @Skcey thanks so much for this contribution! The core architecture looks good and follows the existing patterns well. As a first pass, I'd like to see a few improvements:

1. Replace `fmt.Printf` debug statements with proper structured logging using logr.Logger
2. Fix the serviceResolver.Resolve() calls in TCP/UDP upstream functions to include the missing logger parameter
3. Add comprehensive unit test coverage for all new functions, following existing test patterns
4. As Saylor pointed out above, the RBAC permissions need to be updated and your branch needs to be rebased
5. If possible, could we reduce code duplication between TCP and UDP implementations

Please feel free to reach out with any questions or if you need any help with anything!

Skcey commented:
> @Skcey Can you please rebase this PR? There are a couple changes that went into main that are causing compilation errors.
> 
> Also, the clusterrole in the helm chart needs to be updated to support these routes and their statuses so the controller can actually handle them.

Thanks for the reminder! I’ll keep these issues in mind when I modify the code later. It’s just that I might not have time to update the code for about a week.

Skcey commented:
> Hi @Skcey thanks so much for this contribution! The core architecture looks good and follows the existing patterns well. As a first pass, I'd like to see a few improvements:
> 
> 1. Replace `fmt.Printf` debug statements with proper structured logging using logr.Logger
> 2. Fix the serviceResolver.Resolve() calls in TCP/UDP upstream functions to include the missing logger parameter
> 3. Add comprehensive unit test coverage for all new functions, following existing test patterns
> 4. As Saylor pointed out above, the RBAC permissions need to be updated and your branch needs to be rebased
> 5. If possible, could we reduce code duplication between TCP and UDP implementations
> 
> Please feel free to reach out with any questions or if you need any help with anything!

Thank you so much for your feedback! I will definitely make improvements based on these points. However, I’ll be quite busy over the next week and may not be able to work on this task. I expect to have time to make the changes and refinements the week after next

salonichf5 commented:
> Hi @ciarams87 ! I'm sorry for the long delay. Currently, I have improved the code and implemented the weight-based traffic distribution feature for multiple backend services in TCPRoute and UDPRoute. This has been tested and verified in my own experimental environment. If you have time, could you please help review the code? Thank you very much!

Hi @Skcey,
We’re currently wrapping up the release work for version 2.2, but we’ll be able to shift focus to this very soon.
In the meantime, it seems like the generate files pipeline is failing. When you get a chance, could you please try running the following command from root directory to locally to address the issue?

```
make generate-all
```

It also looks like there are some [linting errors ](https://github.com/nginx/nginx-gateway-fabric/actions/runs/18591970251/job/53010152493?pr=3688#step:5:33)that need fixing. If you could take a look at those as well, that would be great!

Thanks so much for your help, and let us know if you need any support with this.

Skcey commented:
> > Hi @ciarams87 ! I'm sorry for the long delay. Currently, I have improved the code and implemented the weight-based traffic distribution feature for multiple backend services in TCPRoute and UDPRoute. This has been tested and verified in my own experimental environment. If you have time, could you please help review the code? Thank you very much!
> 
> Hi @Skcey, We’re currently wrapping up the release work for version 2.2, but we’ll be able to shift focus to this very soon. In the meantime, it seems like the generate files pipeline is failing. When you get a chance, could you please try running the following command from root directory to locally to address the issue?
> 
> ```
> make generate-all
> ```
> 
> It also looks like there are some [linting errors ](https://github.com/nginx/nginx-gateway-fabric/actions/runs/18591970251/job/53010152493?pr=3688#step:5:33)that need fixing. If you could take a look at those as well, that would be great!
> 
> Thanks so much for your help, and let us know if you need any support with this.

OK! I’ve fixed the issues: update generated files, and resolved all the linting errors.
Let me know if the CI checks pass now, or if anything else is needed!
Thanks so much!

ciarams87 commented on internal/controller/state/graph/route_common.go:
```diff
@@ -1150,3 +1233,168 @@ func routeKeyForKind(kind v1.Kind, nsname types.NamespacedName) RouteKey {
+	var conds []conditions.Condition
+	for _, parentRef := range parentRefs {
+		if err := verifyIPFamily(parentRef.Gateway.EffectiveNginxProxy, svcIPFamily); err != nil {
+			backendRef.Valid = backendRef.Valid || false
```
This is confusing and looks incorrect. Shouldn't it just be 
```suggestion
			backendRef.Valid = false
```
?
